src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -561,16 +567,32 @@ void ObjectMonitor::enter_with_contention_mark(JavaThread *current, ObjectMonito
 
 ObjectMonitor::TryLockResult ObjectMonitor::TryLock(JavaThread* current) {
   void* own = owner_raw();
-  if (own != nullptr) return TryLockResult::HasOwner;
-  if (try_set_owner_from(nullptr, current) == nullptr) {
-    assert(_recursions == 0, "invariant");
-    return TryLockResult::Success;
+  void* first_own = own;
+
+  for (;;) {
+    if (own == DEFLATER_MARKER) {
+      if (TryLockI(current)) {
```

Comment:
I can't tell the difference between TryLockI and enter_for().  Did I previously object to enter_for() here?  Maybe I should take that back, and there should be a comment above enter_for() like
     // Enters a lock in behalf of a non-current thread, or a thread that is exiting and has previously given up the lock.
     // and it handles deflation.

You could add a boolean that you expect success for the enter_for() caller from deoptimization (ie. must_succeed).

This code is getting repetitive - it looks the same in all these places only a little bit different and hard to know why.

Reply:
I'd rather have a `TryLock_with_contention_mark` function which `enter_for` uses. 

Have stronger asserts in `enter_for`, essentially that it is used correctly from Deopt and that the result of `TryLock_with_contention_mark` is true.
```suggestion
      // Block out deflation as soon as possible.
      ObjectMonitorContentionMark contention_mark(this);

      // Check for deflation.
      if (enter_is_async_deflating()) {
        // Treat deflation as interference
        return TryLockResult::Interference;
      }
      if (TryLock_with_contention_mark(current, contention_mark)) {
```

Reply:
Yes @coleenp, you did previously object to calling `enter_for()` from `TryLock()`,  which is why it is what it is today.
I'm not too proud of how it turned out, and as @dholmes-ora also pointed out , the naming is a bit confusing, so that needs to be fixed.

Reply:
I was actually confused because there's an enter_for() in all of the synchronizer files and didn't realize you were calling the one in ObjectMonitor.

Reply:
fixed